torch_geometric/transforms/pad.py

```diff
@@ -0,0 +1,548 @@
+        """
+        if data.is_node_attr(key):
+            self._attr_cache['node'].add(key)
+        elif data.is_edge_attr(key):
+            self._attr_cache['edge'].add(key)
+        else:
+            self._attr_cache['other'].add(key)
+
+    def is_node_attr(self, data: Data, key: str) -> bool:
+        r"""Lookup for :obj:`data.is_node_attr(key)` using cache."""
+        if key in self._attr_cache['node']:
+            return True
+
+        if key in self._attr_cache['edge'] or key in self._attr_cache['other']:
+            return False
+
+        self._put_key(data, key)
+        return self.is_node_attr(data, key)
+
+    def is_edge_attr(self, data: Data, key: str) -> bool:
+        r"""Lookup for :obj:`data.is_edge_attr(key)` using cache."""
+        if key in self._attr_cache['edge']:
+            return True
+
+        if key in self._attr_cache['node'] or key in self._attr_cache['other']:
+            return False
+
+        self._put_key(data, key)
+        return self.is_edge_attr(data, key)
+
+
+@functional_transform('pad')
+class Pad(BaseTransform):
+    r"""Applies padding to enforce consistent tensor shapes
+    (functional name: :obj:`pad`).
+
+    This transform will pad node and edge features up to a maximum allowed
+    size in the node or edge feature dimension. By default :obj:`0.0` is used
+    as the padding value and can be configured by setting :obj:`node_pad_value`
+    and :obj:`edge_pad_value`.
+
+    Note that in case of applying Pad to
+    :class:`~HeteroData` input, the :obj:`node_pad_value`
+    or :obj:`edge_pad_value` can be a single float or a Dict[<node type>,
+    <pad value>]. In the latter case, if the user does not include a particular
+    node type in the dict, this node type's attributes will not get padded. In
+    contrast, if the user does not include a particular edge type in the dict,
+    for which the destination node padding was specified, the default edge
+    padding value is used.
+
+    Note that in order to allow for at least one padding node for any padding
+    edge, below conditions must be met:
+
+    * if :obj:`max_num_nodes` is a single value, it must be greater than the
+      maximum number of nodes of any graph in the dataset;
+    * if :obj:`max_num_nodes` is a dictionary, value for every node type must
+      be greater than the maximum number of this type nodes of any graph in the
+      dataset.
+
+    Args:
+        max_num_nodes (int or Dict[str, int]): The maximum number of nodes.
+            In heterogeneous graphs, may also take in a dictionary denoting the
+            maximum number of nodes for specific node types.
+        max_num_edges (int or Dict[Tuple[str, str, str], int], optional):
+            The maximum number of edges.
+            In heterogeneous graphs, may also take in a dictionary denoting the
+            maximum number of edges for specific edge types.
+        node_pad_value (float or Dict[str, float], optional): The fill value to
```

Comment:
Is there a use-case for passing a dictionary here? Would prefer to keep this simple and just expect a `float`.

Reply:
I believe the request to handle it per node type in case of `HeteroData` came came from our research team. @hatemhelal, could you please help providing a use-case for that?

Reply:
The per-node padding value comes up in cases where I wanted to pad two different node features that have different meanings.  For small molecules:

* `z`: vector of atomic numbers -> use zero as a padding value to represent fake atoms 
* `pos`: 3d atomic position data -> use a large value to exclude padding atoms from any pairwise interactions. this is necessary to avoid wrong answers with padding.